pkg/targetallocator/adapters/config_to_prom_config.go

```diff
@@ -59,6 +78,201 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
+	}
+
+	scrapeConfigs, ok := scrapeConfigsProperty.([]interface{})
+	if !ok {
+		return nil, errorNotAList("scrape_configs")
+	}
+
+	for i, config := range scrapeConfigs {
+		scrapeConfig, ok := config.(map[interface{}]interface{})
+		if !ok {
+			return nil, errorNotAMapAtIndex("scrape_config", i)
+		}
+
+		relabelConfigsProperty, ok := scrapeConfig["relabel_configs"]
+		if !ok {
+			continue
+		}
+
+		relabelConfigs, ok := relabelConfigsProperty.([]interface{})
+		if !ok {
+			return nil, errorNotAListAtIndex("relabel_configs", i)
+		}
+
+		for i, rc := range relabelConfigs {
+			relabelConfig, rcErr := rc.(map[interface{}]interface{})
+			if !rcErr {
+				return nil, errorNotAMapAtIndex("relabel_config", i)
+			}
+
+			replacementProperty, rcErr := relabelConfig["replacement"]
+			if !rcErr {
+				continue
+			}
+
+			replacement, rcErr := replacementProperty.(string)
+			if !rcErr {
+				return nil, errorNotAStringAtIndex("replacement", i)
+			}
+
+			relabelConfig["replacement"] = strings.ReplaceAll(replacement, "$$", "$")
+		}
+
+		metricRelabelConfigsProperty, ok := scrapeConfig["metric_relabel_configs"]
+		if !ok {
+			continue
+		}
+
+		metricRelabelConfigs, ok := metricRelabelConfigsProperty.([]interface{})
+		if !ok {
+			return nil, errorNotAListAtIndex("metric_relabel_configs", i)
+		}
+
+		for i, rc := range metricRelabelConfigs {
+			relabelConfig, ok := rc.(map[interface{}]interface{})
+			if !ok {
+				return nil, errorNotAMapAtIndex("metric_relabel_config", i)
+			}
+
+			replacementProperty, ok := relabelConfig["replacement"]
+			if !ok {
+				continue
+			}
+
+			replacement, ok := replacementProperty.(string)
+			if !ok {
+				return nil, errorNotAStringAtIndex("replacement", i)
+			}
+
+			relabelConfig["replacement"] = strings.ReplaceAll(replacement, "$$", "$")
+		}
+	}
+
+	return prometheus, nil
+}
+
+// AddHTTPSDConfigToPromConfig adds HTTP SD (Service Discovery) configuration to the Prometheus configuration.
+// This function removes any existing service discovery configurations (e.g., `sd_configs`, `dns_sd_configs`, `file_sd_configs`, etc.)
+// from the `scrape_configs` section and adds a single `http_sd_configs` configuration.
+// The `http_sd_configs` points to the TA (Target Allocator) endpoint that provides the list of targets for the given job.
+func AddHTTPSDConfigToPromConfig(prometheus map[interface{}]interface{}, taServiceName string) (map[interface{}]interface{}, error) {
+	prometheusConfigProperty, ok := prometheus["config"]
+	if !ok {
+		return nil, errorNoComponent("prometheusConfig")
+	}
+
+	prometheusConfig, ok := prometheusConfigProperty.(map[interface{}]interface{})
+	if !ok {
+		return nil, errorNotAMap("prometheusConfig")
+	}
+
+	scrapeConfigsProperty, ok := prometheusConfig["scrape_configs"]
+	if !ok {
+		return nil, errorNoComponent("scrape_configs")
+	}
+
+	scrapeConfigs, ok := scrapeConfigsProperty.([]interface{})
+	if !ok {
+		return nil, errorNotAList("scrape_configs")
+	}
```

Comment:
Should we error if the Prometheus config doesn't exist, or just return quietly? See my comment for the TA config unescaping as well.